File: captainhcg/go/35-search-insert-position.go
func searchInsert(nums []int, target int) int {
    if len(nums) == 0{
        return 0
    }
    pos := helper(nums, 0, len(nums)-1, target)
    if nums[pos] == target{
        return pos
    } else if nums[pos] < target{
        return pos + 1
    } else{
        return pos
    }
    
}

func helper(nums []int, start, end, target int) int{
    if start == end{
        return start
    }
    mid := (start + end) / 2
    if nums[mid] == target{
        return mid
    } else if nums[mid] > target{
        return helper(nums, start, mid, target)
    } else {
        return helper(nums, mid+1, end, target)
    }
}

File: bwu/binary_search/35-search-insert-position.py
class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        if n == 0:
            return 0
            
        l, r = 0, n - 1
        while l + 1 < r:
            mid = (l + r) / 2
            if target == nums[mid]:
                return mid
            elif target > nums[mid]:
                l = mid
            else:
                r = mid
                
        if nums[r] < target:
            return r + 1
        elif nums[l] < target:
            return r
        else:
            return l
